jobbergate-api/jobbergate_api/storage.py

```diff
@@ -3,56 +3,177 @@
+    asynchronous: bool = True,
+) -> str:
     """
     Build a database url based on settings.
 
-    If the ``DEPLOY_ENV`` setting is "TEST" or if the ``force_test`` is passed, build from the test database
-    settings.
+    If ``force_test`` is set, build from the test database settings.
+    If ``asynchronous`` is set, use asyncpg.
+    If ``override_db_name`` replace the database name in the settings with the supplied value.
     """
-    is_test = force_test or settings.DEPLOY_ENV.lower() == "test"
-    prefix = "TEST_" if is_test else ""
+    prefix = "TEST_" if force_test else ""
+    db_user = getattr(settings, f"{prefix}DATABASE_USER")
+    db_password = getattr(settings, f"{prefix}DATABASE_PSWD")
+    db_host = getattr(settings, f"{prefix}DATABASE_HOST")
+    db_port = getattr(settings, f"{prefix}DATABASE_PORT")
+    db_name = getattr(settings, f"{prefix}DATABASE_NAME") if override_db_name is None else override_db_name
+    db_path = "/{}".format(db_name)
+    db_scheme = "postgresql+asyncpg" if asynchronous else "postgresql"
 
     return str(
         URL.build(
-            scheme="postgresql",
-            user=getattr(settings, f"{prefix}DATABASE_USER"),
-            password=getattr(settings, f"{prefix}DATABASE_PSWD"),
-            host=getattr(settings, f"{prefix}DATABASE_HOST"),
-            port=getattr(settings, f"{prefix}DATABASE_PORT"),
-            path="/{}".format(getattr(settings, f"{prefix}DATABASE_NAME")),
+            scheme=db_scheme,
+            user=db_user,
+            password=db_password,
+            host=db_host,
+            port=db_port,
+            path=db_path,
         )
     )
 
 
-database = databases.Database(build_db_url(), force_rollback=settings.DEPLOY_ENV.lower() == "test")
+class EngineFactory:
+    """
+    Provide a factory class that creates engines and keeps track of them in an engine mapping.
+
+    This is used for multi-tenancy and database URL creation at request time.
+    """
+
+    engine_map: Dict[str, AsyncEngine]
+
+    def __init__(self):
+        """
+        Initialize the EngineFactory.
+        """
+        self.engine_map = dict()
+
+    async def cleanup(self):
+        """
+        Close all engines stored in the engine map.
+        """
+        for engine in self.engine_map.values():
+            await engine.dispose()
+
+    def get_engine(self, override_db_name: typing.Optional[str] = None) -> AsyncEngine:
+        """
+        Get a database engine.
+
+        If the database url is already in the engine map, return the engine stored there. Otherwise, build
+        a new one, store it, and return the new engine.
+        """
+        db_url = build_db_url(
+            override_db_name=override_db_name,
+            force_test=settings.DEPLOY_ENV.lower() == "test",
+        )
+        if db_url not in self.engine_map:
+            self.engine_map[db_url] = create_async_engine(db_url, pool_pre_ping=True)
+        return self.engine_map[db_url]
```

Comment:
I think we can avoid building the URL every time if we store the engine using `override_db_name`
```suggestion
        if db_url not in self.engine_map:
            db_url = build_db_url(
                override_db_name=override_db_name,
                 orce_test=settings.DEPLOY_ENV.lower() == "test",
            )
            self.engine_map[override_db_name] = create_async_engine(db_url, pool_pre_ping=True)
        return self.engine_map[override_db_name]
```

What do you think?

Reply:
Another approach could be

```suggestion
        return self.engine_map.setdefault(
            override_db_name,
            create_async_engine(
                build_db_url(
                    override_db_name=override_db_name,
                    force_test=settings.DEPLOY_ENV.lower() == "test",
                ),
                pool_pre_ping=True,
            ),
        )
```

Reply:
Well, it's trickier because the `build_db_url` method pulls from settings when the `override_db_name` is None. I didn't really like one of the entries in the map potentially being keyed by `None`. Additionally, I found it handy for introspection to see the full URL for each engine.

As for the `setdefault` approach, I also considered this. However, the arguments aren't lazily evaluated, so the `create_async_engine` method would be called every single time even if there was already an engine for that key.

Reply:
Good point, I see no problem in keeping the code as it is